### ISSUE TITLE
make it easier to share images between the host daemon and a dind

### ISSUE THREAD

**Issue by vikstrous** [moby/moby#29764]
Right now I'm bind mounting `/var/lib/docker/overlay2` and `/var/lib/docker/images/overlay2` and it allows me to share images between the host daemon and the dinds I'm running. This approach has some limitations though. If I pull in one daemon, it seems like the rest of the daemons get confused and I can't use the image I pulled in them. I can fix this by restarting the other daemons. It would be really useful to be able to do this kind of image sharing without hacks.

**Comment by thaJeztah** [comment]
This is really dangerous stuff; `/var/lib/docker` is designed to be accessed by a single daemon; having multiple daemons access the same directory is asking for trouble

https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

**Comment by vikstrous** [comment]
Yep. Sometimes you gotta live on the edge :P

Our use case is testing DTR+UCP clusters.

We can test them on VMs but there are a few major limitations:

1. There's only so much RAM on a dev laptop, so in the best case maybe you can run 3 VMs if you close chrome, but you'll never be able to run more than one cluster at a time.
2. It takes a long time to to transfer images from the host daemon to the VMs
3. Creating and destroying VMs is slow
4. It requires a lot of manual and careful operations to keep your VMs in a good state
5. The VM management requires lots and lots of bash aliases and functions to make it bearable and they are custom for each developer
6. Our CI can't run VMs the same way because it is itself running on a VM so instead it used terraform to provision AWS VMs, which is an entirely different code path
7. Mac users can't use automated KVM VM management tools if we start building things that way and supporting a number of hypervisors and somehow escaping the dev VM to control them is a project I really don't want to get into

For these reasons we switched to DIND for CI. This solved the above problems:

1. It uses way less RAM
2. Image transfers are unnecessary
3. It's fast and easy to run dinds
4. Because of how fast it is, we can just blow away the whole environment and start from scratch every time pretty quickly without having to be careful about the state of VMs
5. We have a tool that runs docker commands and it works for everyone because docker is more standard than hypervisors (we use linux and mac so everyone has their own favorite hypervisor: KVM, VirtualBox, VMWare, etc.)
6. Our CI runs the same thing as dev. We haven't fully replaced everything with DIND though. Our nightly builds still run on real machines because DIND will never be able to test everything.
7. We don't need to do any VM hacks.

Basically, we are already using DIND for CI very successfully and we just want to get it a bit more polished. The main pain right now is around image sharing.

For now we re-create the dind whenever we change images and we pull only from the host daemon. I'll look into using image tars for transferring images every time we run tests, but that would not be ideal.

**Comment by thaJeztah** [comment]
Why not use named volumes for the dind containers, so that they can persist beyond the containers lifecycle, but still allows you to have a "private" storage per daemon?

    -v dind-01:/var/lib/docker

**Comment by vikstrous** [comment]
The persistence of the images is not the problem. The images are built using the host daemon and then need to be transferred or reused in the dinds. Right now I see the following options:

- bind mount the host daemon's storage and careful not to pull anything
- keep the dind daemons' storage separate and transfer the images via tars
- keep the dind daemons' storage separate and transfer the images by pushing and pulling from a registry

**Comment by vdemeester** [comment]
@vikstrous I would think that the "easiest" setup would be 3. for CI and devs too.

**Comment by vikstrous** [comment]
Yeah and that's what I used until I spent a month in Africa. Even in Canada a residential connection is way too slow for pushing and pulling 200MB+ images between builds. Residential connections tend to be asymmetric and often have less than 1 mbps upload speed.

**Comment by thaJeztah** [comment]
@vikstrous you can run that registry locally, or even as a service;

    docker service create --name registry --publish 5000:5000 registry:2

then use `127.0.0.1:5000` as your registry

**Comment by vikstrous** [comment]
(For anyone reading this: `--publish 127.0.0.1:5000:5000` would be safer if run directly on a laptop. You wouldn't want to accidentally expose dev images to the local network.)

DTR images have their namespace hardcoded in them so that when you run bootstrapper commands, the bootstrapper knows where to find the DTR images. Therefore to use a local registry I have to:

1. Run a local registry
2. Modify the image names in our build system
3. Build special images that use the local registry's name
4. Push+pull, etc.

It can be done and it can be automated, but it seems much more difficult than option 2. If the DTR bootstrapper knew its own image name, it would be possible to avoid the image name hardcoding and then option 3 would be much more sensible. One way to do that might be to expose the image name to the container through something like this: https://github.com/docker/docker/pull/26331

Option 2 is basically `docker save image1 image2 image3 | docker exec -i dind1 docker load`, so it's very appealing.

Anyway, here are some actual experiments.

|                                 | option 1 | option 2 | option 3 |
|----------------------------|------------|-------------|------------|
| speed from scratch | t(dind recreate) + t(ucp install) | 1 minute | 1 minute push + 1 minute pull |
| speed for re-transfer of same build | t(dind recreate) + t(ucp install) | 20 seconds | 1 sec push + 1 sec pull |
| speed for transfer of incremental build | t(dind recreate) + t(ucp install) | 1 minute (?!) | 40 sec push +35 sec pull | 

I was surprised to find out that it's not any faster to transfer images that share most of the same base layers via save/load compared to completely new images. Looking at `docker history` it seems like shared base images are not preserved between `save/load`.

t(dind recreate) is about 10 seconds, and t(ucp install) is about 30 seconds. Even re-creating the dind container and installing UCP on it is faster than image transfer using methods 2 or 3.

* Our base image is 450 MB uncompressed
* We have 9 derived images with around 70-140 MB on top of the base image
* We have another 30 MB image
* And another 70 MB image

With smaller images it might make sense to do the image transfer thing, but at this size, it seems like options 2 and 3 are not reasonable for interactive dev work and option 3 is almost always slower than option 2 even though save/load is not optimized at all.

I'll stick with option 1 for now, but I'm hoping that more can be done to speed things up even further by either not requiring a dind restart or making it possible to restart dind without re-creating it.

(Right now we are forced to re-create the dind because if we restart it, networking breaks in strange and racy ways. That bug deserves its own issue, but I don't have a minimal repro for it yet.)

**Comment by bsousaa** [comment]
It looks like this issue went stale. Let me close it for now. Feel free to comment and we can re-open if we have more info.
